odoo/addons/base/models/ir_attachment.py

```diff
@@ -563,70 +573,92 @@ def _search(self, domain, offset=0, limit=None, order=None, *, active_test=True,
-            if res_model not in self.env:
-                allowed_ids.update(id_ for ids in targets.values() for id_ in ids)
+        domain = domain.optimize(self)
+        if self.env.su or bypass_access or domain.is_false():
+            return super()._search(domain, offset, limit, order, active_test=active_test, bypass_access=bypass_access)
+
+        # General access rules
+        # - public == True are always accessible
+        sec_domain = Domain('public', '=', True)
+        # - res_id == False needs to be system user or creator
+        res_ids = condition_on_field(self, 'res_id', domain)
+        if not res_ids or False in res_ids:
+            if self.env.is_system():
+                sec_domain |= Domain('res_id', '=', False)
+            else:
+                sec_domain |= Domain('res_id', '=', False) & Domain('create_uid', '=', self.env.uid)
+
+        # Searching for all documents or a subset of models
+        res_model_names = condition_on_field(self, 'res_model', domain)
+        if not (0 < len(res_model_names) <= 5):
+            # We do not have a small restriction on res_model. We still need to
+            # support other queries such as: `('id', 'in' ...)`.
+            # Always limit the search so that we avoid reading the whole table.
+            # If the limit is checked and reached, return no results.
+            check_limit = False
+            if limit is None:
+                if ids := condition_on_field(self, 'id', domain):
+                    limit = len(ids)
+                else:
+                    limit = 20 if modules.module.current_test else PREFETCH_MAX
+                    check_limit = True
+            # Restrict with domain and add all attachments linked to res_model.
+            domain &= sec_domain | Domain('res_model', '!=', False)
+            domain = domain.optimize_full(self)
+            result = []
+            sub_offset = 0
+            limit += offset
+            while len(result) < limit:
+                records = self.sudo().with_context(active_test=False).search_fetch(
+                    domain,
+                    SECURITY_FIELDS,
+                    offset=sub_offset,
+                    limit=PREFETCH_MAX,
+                    # by default, order by model to batch access checks
+                    order=order or 'res_model nulls first, id',
+                ).sudo(False)
+                result.extend(records._filtered_access('read')._ids)
+                if len(records) < PREFETCH_MAX:
+                    # There are no more records
+                    break
+                if check_limit and sub_offset > limit:
+                    # Search is not restrictive enough, bail out.
+                    return self.browse()._as_query()
+                sub_offset += PREFETCH_MAX
+            if check_limit and len(result) >= limit:
+                # Too many attachments found, just return nothing.
+                return self.browse()._as_query()
```

Comment:
Raise or something else, silently return nothing is weird.

Reply:
Split into a separate commit. Not sure if we should keep it.